app/models/concerns/orchestration/compute.rb

```diff
@@ -102,19 +103,51 @@ def delUserData
   def setComputeDetails
     if vm
       attrs = compute_resource.provided_attributes
-      normalize_addresses if attrs.keys.include?(:mac) or attrs.keys.include?(:ip)
+#      normalize_addresses if attrs.keys.include?(:mac) or attrs.keys.include?(:ip)
+
+      # mac and ip are properties of the NIC, and there may be more than one,
+      # so we need to loop. First store the nics returned from Fog in a local
+      # array so we can delete from it
+      fog_nics = vm.interfaces.dup
 
       attrs.each do |foreman_attr, fog_attr |
-        # we can't ensure uniqueness of #foreman_attr using normal rails validations as that gets in a later step in the process
-        # therefore we must validate its not used already in our db.
-        value = vm.send(fog_attr)
-        value ||= find_address if foreman_attr == :ip
-        self.send("#{foreman_attr}=", value)
-
-        if value.blank? or (other_host = Host.send("find_by_#{foreman_attr}", value))
-          delCompute
-          return failure("#{foreman_attr} #{value} is already used by #{other_host}") if other_host
-          return failure("#{foreman_attr} value is blank!")
+        if foreman_attr == :mac
+          #TODO, do we need handle :ip as well? for openstack / ec2 we only set a single
+          # interface (so host.ip will be fine), and we'd need to rethink #find_address :/
+
+          self.interfaces.each do |nic|
+            selected_nic = vm.select_nic(fog_nics, nic)
+            next if selected_nic.nil? # found no matching fog nic for this Foreman nic, move on
+            value   = selected_nic.send(fog_attr)
+            logger.debug "Orchestration::Compute: nic #{nic.inspect} assigned to #{selected_nic.inspect}"
+            nic.send("#{foreman_attr}=",value)
+            fog_nics.delete(selected_nic) # don't use the same fog nic twice
+
+            # we can't ensure uniqueness of #foreman_attr using normal rails
+            # validations as that gets in a later step in the process
+            # therefore we must validate its not used already in our db.
+
+            # In future, we probably want to skip validation of macs/ips on the Nic
+            # macs can be duplicated if we are creating bonds
+            # ips can be duplicated if we have isolated subnets (needs an update in the Subnet model first)
+            # For now, we scope to physical devices only for the validations
+            if value.blank? or (other = Nic::Base.physical.send("find_by_#{foreman_attr}", value))
```

Comment:
This is really confusing to read, the second part of the statement may not assign anything. Also 'value' and 'other' are really vague terms, it took me a while to understand what exactly they were, please consider changing their name to something meaningful in the context.

Is this equivalent to the current code?

```ruby
if value.blank?
  delCompute
  failure("#{foreman_attr} value is blank!")
elsif (other = Nic::Base.physical.send("find_by_#{foreman_attr}", value))
  delCompute
  failure("#{foreman_attr} #{value} is already used by #{other}")
end
```

That ```or ``` here is used for control flow, looks like it was on purpose since you check for the existence of ```other``` a bit later, I think it's rather confusing and can possibly be rewritten in an easier way?
http://devblog.avdi.org/2010/08/02/using-and-and-or-in-ruby/